### PR TITLE
[dagster-dlt] Fix extract_resource_metadata and fetch_row_count_metadata crash when source yields no data

### LINKED ISSUE #33572: [dagster-dlt] extract_resource_metadata crashes with AttributeError when source yields no data
# `[dagster-dlt] extract_resource_metadata crashes with AttributeError when source yields no data`

## Summary

`DagsterDltResource.extract_resource_metadata` crashes when a dlt source returns no data, because `last_normalize_info` is accessed without a `None` guard.

## Bug

In `dagster_dlt/resource.py` ([line 134](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-dlt/dagster_dlt/resource.py#L134)):

```python
rows_loaded = dlt_pipeline.last_trace.last_normalize_info.row_counts.get(
    normalized_table_name
)
```

When a source yields no data, dlt skips the normalize step entirely. `dlt.pipeline.trace.last_normalize_info` returns `None` in this case (see [`dlt/pipeline/trace.py`](https://github.com/dlt-hub/dlt/blob/master/dlt/pipeline/trace.py) — `last_pipeline_step_trace("normalize")` finds no matching step). The chained `.row_counts` access then raises:

```
AttributeError: 'NoneType' object has no attribute 'row_counts'
```

This causes the entire asset materialization to fail even though `dlt_pipeline.run()` and `load_info.raise_on_failed_jobs()` both completed successfully — the crash happens during metadata extraction, not during the actual load.

## Reproduction

1. Create a `@dlt_assets` definition with a source that conditionally yields no data (e.g., an API that returns an empty response for a given time window).
2. Materialize the asset when the source has no data to return.
3. `extract_resource_metadata` raises `AttributeError`.

## Suggested fix

Guard `last_normalize_info` before accessing `.row_counts`:

```python
normalize_info = dlt_pipeline.last_trace.last_normalize_info
rows_loaded = (
    normalize_info.row_counts.get(normalized_table_name)
    if normalize_info is not None
    else None
)
if rows_loaded:
    base_metadata["rows_loaded"] = MetadataValue.int(rows_loaded)
```

## Context

- Introduced in PR #24306 (Sep 2024)
- Affects `dagster-dlt` 0.28.18 and current `master`
- The `.fetch_row_count()` chain on `DltEventIterator` has a similar gap — `fetch_row_count_metadata` raises when `jobs` metadata is an empty list (falsy), which also occurs on no-data runs.

### PR DESCRIPTION
**Pull request by vidiyala99**
Summary & Motivation
Fixes #33572. This PR addresses an AttributeError that occurs when a dlt source yields no data. In such cases, the dlt pipeline skips the normalization step, causing dlt_pipeline.last_trace.last_normalize_info to be None. Attempting to access .row_counts on this None object resulted in a crash.

Additionally, this PR fixes a related gap in fetch_row_count_metadata where an empty or missing jobs metadata list would raise an Exception instead of degrading gracefully (returning None for the row count).

Changes
extract_resource_metadata: Added safety checks to handle cases where last_trace or last_normalize_info are None, ensuring metadata extraction continues without crashing.

fetch_row_count_metadata: Modified to return TableMetadataSet(row_count=None) and log a debug message when the jobs metadata list is empty or missing, matching existing error-logging patterns.

Test Plan
Added three new unit tests to dagster_dlt_tests/test_fetch_metadata.py using mocks to ensure robustness against dlt internal variations without requiring a live connection:

test_fetch_row_count_no_jobs: Verifies that fetch_row_count_metadata returns row_count=None and logs a debug message when the jobs list is empty.

test_extract_resource_metadata_no_normalize_info: Verifies that extract_resource_metadata handles last_normalize_info=None by omitting rows_loaded from metadata.

test_extract_resource_metadata_no_last_trace: A defensive test ensuring the logic also handles the edge case where last_trace itself is None.

Changelog
[dagster-embedded-elt] BUGFIX: Fixed extract_resource_metadata and fetch_row_count_metadata crash when a dlt source yields no data.

### CODE REVIEW

File: python_modules/libraries/dagster-dlt/dagster_dlt_tests/test_fetch_metadata.py [hunk 1]
```diff
@@ -127,3 +127,127 @@ def example_pipeline_assets(
...
+    )
+
+    context = mock.MagicMock()
+    context.has_partition_key = False
+    dlt_pipeline_mock = mock.MagicMock()
+
+    result = fetch_row_count_metadata(materialization, context, dlt_pipeline_mock)
+    assert isinstance(result, TableMetadataSet)
+    assert result.row_count is None
+    context.log.debug.assert_called_once()
```
[the remark this answers is not in the document]
Comment (vidiyala99) [reply python_modules/libraries/dagster-dlt/dagster_dlt_tests/test_fetch_metadata.py:156]: Fixed. I've added the missing assertion for the second test case (missing jobs key) after the mock reset to ensure that both scenarios are fully verified for logging.
